proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+(ref (instance ...)) <: instanceref
+```
+
+The previous section noted that `instance.instantiate` is, in this proposal,
+only allowed in the constant initializer expressions of `instance` definitions.
+The immediate reason for this is the GC requirement stated [above](#additional-requirements):
+since instances can form graphs (via mutable tables and globals), a first-class
+`instance.instantiate` instruction would inherently require a garbage collection
+algorithm to collect dead instances. This could be added later on as part of a
+batch of GC features (viz., [GC types]) with a careful definition of what
+features are available on hosts that don't support GC. But for now, this proposal
+aims to be universally available.
+
+In contrast, `ref.module`/`ref.instance` and module/instance references have no
+such inherent GC requirement and would be useful in allowing hosts to optionally
+expose runtime instantiation through *import APIs*. For example, using
+[Interface Types] to express the signature more succinctly, a classic
+[`dlopen`]/[`dlsym`] interface could be approximated:
+```wasm
+(instance
+  (export "dlopen" (func
+    (param $module moduleref)
+    (param $imports (array (record (field "name" string) (field "value" anyref))))
+    (result instanceref)
+  ))
+  (export "dlsym" (func
+    (param $instance instanceref)
+    (param $symbol string)
+    (result funcref)
+  ))
+)
+```
+
+Thus, as a conservative first step toward runtime dynamic linking, this proposal
+includes `ref.module` and `ref.instance` in the general wasm instruction set and
+includes module and instance reference types in the general [`reftype`] set.
+
+Lastly, since modules can not just import individual functions, but also individual
+memories, tables and globals, new `ref.memory`, `ref.table` and `ref.global`
+instructions and reference types are added:
+```wasm
+ref.memory $memory : [] -> [ (ref (memory ...)) ]
+ref.table $table : [] -> [ (ref (table ...)) ]
+ref.global $global : [] -> [ (ref (global ...)) ]
+```
+(These instructions have been planned since the first Reference Types proposal.)
+
+
+### Nested Modules
+
+Symmetric to nested instances, modules can contain *nested modules* via
+`module` definitions. Nested modules are injected into the same module index
+space as module imports and thus can be instantiated the same way. For example:
+```wasm
+(module
+  (module $child
+    (func $hi (export "hi")
+      ...
+    )
+  )
+  (instance $i (instance.instantiate $child))
+  (func (export "run")
+    call $i.$hi
+  )
+)
+```
+
+Unlike most source-language nested functions/classes, nested modules have no
+special access to their parents' state. However, since modules and types are
+closed, stateless expressions and thus "sharing" is equivalent to duplication,
+nested modules can refer to their parents' types and modules via *aliases*. As
+syntactic sugar in the textual format, a nested module can simply use the
+`$name` of the enclosing module/type:
```

Comment:
I don't think it's accurate to describe this as syntactic sugar. When referring to _imports_ of an enclosing module, they cannot be syntactically expanded. They can only be substituted when the enclosing module is instantiated. So semantically, this really is a lexical closure.

For example:
```
(module $O
  (import "M" (module $M ...))
  (module $N (export (module $M)))
  (export (module $N))
)
```
is the exact analogue to a function returning a closure:
```
func O(M : func ...) {
  func N() { return M };
  return N;
}
```
Unlike with function closures, it may be appropriate to implement this with actual substitution, though. Is that what you're getting at?

Reply:
Technically, I used "syntactic sugar" to refer to the ability to directly use the identifier of an enclosing module's typedef/module, without an explicit `alias` definition; I think that is sugar, but of a different kind.  But probably I should speak a bit more clearly to this nuance.  LMKWYT.

Reply:
Ah okay, I see. That makes sense.